wrc/codegen/merger.py

```diff
@@ -0,0 +1,177 @@
+from wrc.sema.ast import Article, Section, TableOfContent, Guideline, split_rule_number
+from copy import deepcopy
+
+NOTES_INDEX = 0
+TOC_INDEX = 1
+FIRST_ARTICLE_INDEX = 2
+
+
+class BadFormatError(Exception):
+    def __init__(self, message: str):
+        self.message = message + '. Please report this issue to the developer.'
+        super().__init__(message)
+
+
+def compare_article_numbers(num1: str, num2: str) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    This is necessary because:
+    * We need to compare article numbers like '1' and 'A'.
+    * '2' > '10' == True, which is a problem.
+    """
+    is_digit_1 = num1.isdigit()
+    is_digit_2 = num2.isdigit()
+    if is_digit_1 and is_digit_2:
+        res = int(num1) < int(num2)
+    elif is_digit_1 and not is_digit_2:
+        # Only num2 is a letter.
+        res = True
+    elif not is_digit_1 and not is_digit_2:
+        res = num1 < num2
+    else:
+        res = False
+
+    return res
+
+
+def compare_reg_numbers(num1: list, num2: list) -> bool:
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    We need this for the same reasons we need compare_article_numbers().
+    """
+    for n1, n2 in zip(num1, num2):
+        if n1 < n2:
+            return True
+        elif n1 > n2:
+            return False
+
+    return len(num1) < len(num2)
+
+
+def get_reg_number(reg):
+    if isinstance(reg, Guideline):
+        reg = reg.number.split('+')[0]
+        num = split_rule_number(reg)
+    else:
+        num = split_rule_number(reg.number)
+    if str(num[-1]) == '0':
+        # Because split_rule_number() sometimes returns a 0 at the end, and we don't want that here.
+        del num[-1]
+    return num
```

Comment:
Most languages have a `trim_end` function or similar...

Reply:
The alternatives are:
- `num.pop(-1)`, but it also returns the removed value, which is not necessary.
- `num = num[:-1]`, which is not what you are looking for.

I just made a quick search and I don't think there's anything to specifically remove the last element.

Reply:
I might be missing something. `num` is not a string, it's a list that contains both letters and numbers. I'm just leveraging `split_rule_number()` which returns such a list and then I check if we need to remove the last element if it's a 0. I'm converting the last element to a string before making the comparison in order to avoid an exception if the last element is not a number (str == int comparison is not legal), but `num` is still a list. `num` is probably not the happiest name :p